File: v2/131.py





class Solution:
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """
        # 6 star, 必须掌握
        # 辅助函数判断字符串是否回文很重要
        # 从前往后遍历，如果前面某部分是一个回文，将这部分加到结果集中
        rs = []
        self.dfs([], rs, s)
        return rs

    def is_palindrome(self, s):
        n = len(s)
        for i in range(n//2 + 1):
            if s[i] != s[n-1-i]:
                return False
        return True

    def dfs(self, path, rs, s):
        if s == "":
            rs.append(path)
            return
        for i in range(1, len(s)+1):
            if self.is_palindrome(s[:i]):
                self.dfs(path+[s[:i]], rs, s[i:])


print(Solution().partition("aab"))

File: go/131.go

//package main
//
//import "fmt"

func partition(s string) [][]string {
	var ret [][]string
	dfs(s, []string{}, &ret)
	return ret

}


func isPalindrome(s string) bool {
	length := len(s)
	for i:=0; i < length/2 +1; i++ {
		if s[i] != s[length-1-i] {
			return false
		}
	}
	return true

}


func dfs(s string, path []string, ret *[][]string) {
	if s == "" {
		// 注意这里创建了一个新的切片，然后将path拷贝到新的切片中，然后把新的切片放到结果集中
		// 因为如果使用path这个切片，可能对影响到已经放入结果集的切片，因为切片共享一个底层数组
		temp := make([]string, len(path))
		copy(temp, path)
		*ret = append(*ret, temp)
		return
	}
	for i:=1; i < len(s)+1; i++ {
		if isPalindrome(s[:i]) {
			dfs(s[i:], append(path, s[:i]), ret)
		}
	}
}


//func main() {
//	s := "abababbaba"
//	ret := partition(s)
//	//fmt.Println(ret)
//	for i, _ := range ret {
//		fmt.Println(ret[i])
//	}
//}